lib/trino-parquet/src/main/java/io/trino/parquet/reader/ChunkedInputStream.java

```diff
@@ -0,0 +1,151 @@
+import static com.google.common.io.ByteStreams.readFully;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A single continuous {@link InputStream} over multiple {@link Slice}s read on demand using given collection of {@link ChunkReader}s.
+ * It is used to read parquet column chunk in limited (small) byte chunks (8MB by default, controlled by {@link ParquetReaderOptions#getMaxReadBlockSize()}).
+ * Column chunks consists of multiple pages.
+ * This abstraction is used because the page size is unknown until the page header is read
+ * and page header and page data can be split between two or more byte chunks.
+ */
+public final class ChunkedInputStream
+        extends InputStream
+{
+    private final Iterator<? extends ChunkReader> chunks;
+    private ChunkReader currentChunkReader;
+    private BasicSliceInput current;
+
+    public ChunkedInputStream(Collection<? extends ChunkReader> chunks)
+    {
+        requireNonNull(chunks, "chunks is null");
+        checkArgument(!chunks.isEmpty(), "At least one chunk is expected but got none");
+        this.chunks = chunks.iterator();
+        readNextChunk();
+    }
+
+    public Slice getSlice(int length)
+            throws IOException
+    {
+        if (length == 0) {
+            return Slices.EMPTY_SLICE;
+        }
+        ensureOpen();
+        while (!current.isReadable()) {
+            checkArgument(chunks.hasNext(), "Requested %s bytes but 0 was available", length);
+            readNextChunk();
+        }
+        if (current.available() >= length) {
+            return current.readSlice(length);
+        }
+        // requested length crosses the slice boundary
+        byte[] bytes = new byte[length];
+        try {
+            readFully(this, bytes, 0, bytes.length);
+        }
+        catch (IOException e) {
+            throw new RuntimeException("Failed to read " + length + " bytes", e);
+        }
+        return Slices.wrappedBuffer(bytes);
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len)
+            throws IOException
+    {
+        checkPositionIndexes(off, off + len, b.length);
+        if (len == 0) {
+            return 0;
+        }
+        ensureOpen();
+        while (!current.isReadable()) {
+            if (!chunks.hasNext()) {
+                return -1;
+            }
+            readNextChunk();
+        }
+
+        return current.read(b, off, len);
+    }
+
+    @Override
+    public int read()
+            throws IOException
+    {
+        ensureOpen();
+        while (!current.isReadable() && chunks.hasNext()) {
+            readNextChunk();
+        }
+
+        return current.read();
+    }
+
+    @Override
+    public int available()
+            throws IOException
+    {
+        ensureOpen();
+        return current.available();
+    }
+
+    @Override
+    public void close()
+    {
+        if (current == null) {
+            // already closed
+            return;
+        }
+        currentChunkReader.free();
```

Comment:
If this method throws exception, we will not close chunks – this is expected behavior?

Reply:
`free` does not throw any exceptions, but even if it did, IMO we should just fail the query here, and freeing other chunks doesn't matter.